File: backend/knowledge/models/knowledge.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from sqlalchemy import Integer, String, Text, Float, DateTime, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class KnowledgeCollection(Base):
    __tablename__ = "knowledge_collections"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    project_id: Mapped[Optional[str]] = mapped_column(String(128), nullable=True)
    name: Mapped[str] = mapped_column(String(256), nullable=False)
    domain: Mapped[str] = mapped_column(String(64), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    default_language: Mapped[Optional[str]] = mapped_column(String(16), nullable=True)
    extra_metadata: Mapped[Optional[Any]] = mapped_column("metadata", JSONB, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )


class KnowledgeDocument(Base):
    __tablename__ = "knowledge_documents"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    collection_id: Mapped[int] = mapped_column(Integer, nullable=False)
    external_id: Mapped[Optional[str]] = mapped_column(String(128), nullable=True)
    title: Mapped[str] = mapped_column(String(512), nullable=False)
    source_uri: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    extra_metadata: Mapped[Optional[Any]] = mapped_column("metadata", JSONB, nullable=True)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="raw")

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )


class KnowledgeChunk(Base):
    __tablename__ = "knowledge_chunks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    document_id: Mapped[int] = mapped_column(Integer, nullable=False)
    chunk_index: Mapped[int] = mapped_column(Integer, nullable=False)

    section_label: Mapped[Optional[str]] = mapped_column(String(256), nullable=True)
    text: Mapped[str] = mapped_column(Text, nullable=False)

    tags: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    embedding: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    importance: Mapped[float] = mapped_column(Float, nullable=False, default=0.5)
    extra_metadata: Mapped[Optional[Any]] = mapped_column("metadata", JSONB, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

File: frontend/src/api/knowledge.ts
import { http } from './http'

export type KnowledgeCollection = {
  id: number
  project_id: string | null
  name: string
  domain: string
  description?: string | null
  default_language?: string | null
  extra_metadata?: unknown
  created_at: string
  updated_at: string
}

export type KnowledgeDocument = {
  id: number
  collection_id: number
  external_id?: string | null
  title: string
  source_uri?: string | null
  extra_metadata?: unknown
  status: string
  created_at: string
  updated_at: string
}

export type KnowledgeChunk = {
  id: number
  document_id: number
  chunk_index: number
  section_label?: string | null
  text: string
  tags?: unknown
  importance?: number | null
  metadata?: unknown
}

export type CreateCollectionPayload = {
  project_id?: string
  name: string
  domain: string
  description?: string
  default_language?: string
  metadata?: unknown
}

export type CreateDocumentPayload = {
  collection_id: number
  external_id?: string
  title: string
  source_uri?: string
  metadata?: unknown
}

export async function listCollections(params?: {
  project_id?: string
  domain?: string
}): Promise<{ items: KnowledgeCollection[] }> {
  const query = new URLSearchParams()
  if (params?.project_id) query.set('project_id', params.project_id)
  if (params?.domain) query.set('domain', params.domain)
  const qs = query.toString()
  const url = '/api/knowledge/collections' + (qs ? `?${qs}` : '')
  return http.get<{ items: KnowledgeCollection[] }>(url)
}

export async function createCollection(
  payload: CreateCollectionPayload,
): Promise<KnowledgeCollection> {
  return http.post<KnowledgeCollection>('/api/knowledge/collections', payload)
}

export async function listDocuments(params?: {
  collection_id?: number
  status?: string
}): Promise<{ items: KnowledgeDocument[] }> {
  const query = new URLSearchParams()
  if (params?.collection_id) query.set('collection_id', String(params.collection_id))
  if (params?.status) query.set('status', params.status)
  const qs = query.toString()
  const url = '/api/knowledge/documents' + (qs ? `?${qs}` : '')
  return http.get<{ items: KnowledgeDocument[] }>(url)
}

export async function createDocument(
  payload: CreateDocumentPayload,
): Promise<KnowledgeDocument> {
  return http.post<KnowledgeDocument>('/api/knowledge/documents', payload)
}

export async function deleteDocument(documentId: number): Promise<{
  document_id: number
  deleted_chunks: number
  status: string
}> {
  return http.delete<{ document_id: number; deleted_chunks: number; status: string }>(
    `/api/knowledge/documents/${documentId}`,
  )
}

export async function indexDocument(documentId: number): Promise<{
  document_id: number
  chunk_count: number
  status: string
}> {
  return http.post<{ document_id: number; chunk_count: number; status: string }>(
    `/api/knowledge/documents/${documentId}/index`,
    {},
  )
}

export async function listChunks(documentId: number): Promise<{
  items: KnowledgeChunk[]
}> {
  return http.get<{ items: KnowledgeChunk[] }>(`/api/knowledge/documents/${documentId}/chunks`)
}

export async function resetCollection(collectionId: number): Promise<{
  collection_id: number
  deleted_documents: number
  deleted_chunks: number
  status: string
}> {
  return http.post<{
    collection_id: number
    deleted_documents: number
    deleted_chunks: number
    status: string
  }>(`/api/knowledge/collections/${collectionId}/reset`, {})
}
